codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java

```diff
@@ -300,74 +291,108 @@ public static String decodeComponent(final String s, final Charset charset) {
+
+        CharsetDecoder decoder = CharsetUtil.decoder(charset);
+        int decodedCapacity = len / 3;
+        ByteBuffer byteBuf = ByteBuffer.allocate(decodedCapacity);
+        CharBuffer charBuf = CharBuffer.allocate(decodedCapacity);
+
+        StringBuilder strBuf = new StringBuilder(len);
+        strBuf.append(s, from, firstEscaped);
+
+        for (int i = firstEscaped; i < toExcluded; i++) {
             char c = s.charAt(i);
-            switch (c) {
-                case '+':
-                    buf[pos++] = ' ';  // "+" -> " "
-                    break;
-                case '%':
-                    if (i == size - 1) {
-                        throw new IllegalArgumentException("unterminated escape"
-                                + " sequence at end of string: " + s);
-                    }
-                    c = s.charAt(++i);
-                    if (c == '%') {
-                        buf[pos++] = '%';  // "%%" -> "%"
-                        break;
-                    }
-                    if (i == size - 1) {
-                        throw new IllegalArgumentException("partial escape"
-                                + " sequence at end of string: " + s);
-                    }
-                    c = decodeHexNibble(c);
-                    final char c2 = decodeHexNibble(s.charAt(++i));
-                    if (c == Character.MAX_VALUE || c2 == Character.MAX_VALUE) {
-                        throw new IllegalArgumentException(
-                                "invalid escape sequence `%" + s.charAt(i - 1)
-                                + s.charAt(i) + "' at index " + (i - 2)
-                                + " of: " + s);
-                    }
-                    c = (char) (c * 16 + c2);
-                    // Fall through.
-                default:
-                    buf[pos++] = (byte) c;
-                    break;
+            if (c != '%') {
+                strBuf.append(c != '+' || isPath? c : SPACE);
+                continue;
+            }
+
+            byteBuf.clear();
+            do {
+                byteBuf.put(decodeHexByte(s, i, toExcluded));
+                i += 3;
+            } while (i < toExcluded && s.charAt(i) == '%');
+            i--;
+
+            byteBuf.flip();
+            charBuf.clear();
+            CoderResult result = decoder.reset().decode(byteBuf, charBuf, true);
+            try {
+                if (!result.isUnderflow()) {
+                    result.throwException();
+                }
+                result = decoder.flush(charBuf);
+                if (!result.isUnderflow()) {
+                    result.throwException();
+                }
+            } catch (CharacterCodingException ex) {
+                throw new IllegalStateException(ex);
             }
+            strBuf.append(charBuf.flip());
+        }
+        return strBuf.toString();
+    }
+
+    private static byte decodeHexByte(String s, int pos, int len) {
```

Comment:
doesn't have to be addressed as part of this PR ... I meant to create an issue but got tracked. https://github.com/netty/netty/issues/6797